### PR TITLE
fix(mutate): do not lift table column that results from mutate

### LINKED ISSUE #3515: bug: clickhouse and impala semi/anti join code is wrong
Impala and clickhouse are generating incorrect code in the tests from https://github.com/ibis-project/ibis/pull/3501.

### LINKED ISSUE #3529: bug: sqlalchemy backends create unbound anon column when projecting a mutated column
A join with `batting` and `awards_players` works as expected:

    con = ibis.sqlite.connect("ci/ibis-testing-data/ibis_testing.db")
    batting = con.table("batting")
    awards_players = con.table("awards_players")
    awards_players
    left = batting
    right = awards_players
    expr = left.join(right, "yearID")
    expr.limit(10).execute()

If I `mutate` to rename a column in `batting` things still work:

    awards_players = con.table("awards_players")
    batting = con.table("batting")
    left = batting.mutate(year=batting.yearID)
    right = awards_players
    expr = left.join(right, left.year == right.yearID)
    expr.limit(10).execute()

But if I then create a projection with the newly created column it goes badly:

    awards_players = con.table("awards_players")
    batting = con.table("batting")
    left = batting.mutate(year=batting.yearID)
    left = left['year', 'RBI']
    right = awards_players
    expr = left.join(right, left.year == right.yearID)
    expr.limit(10).execute()

Fails with

    OperationalError: (sqlite3.OperationalError) no such column: anon_1.year
    [SQL: SELECT t0.year, t0."RBI", t1."playerID", t1."awardID", t1."yearID", t1."lgID", t1.tie, t1.notes
    FROM (SELECT t2.year AS year, t2."RBI" AS "RBI"
    FROM (SELECT t3."playerID" AS "playerID", t3."yearID" AS "yearID", t3.stint AS stint, t3."teamID" AS "teamID", t3."lgID" AS "lgID", t3."G" AS "G", t3."AB" AS "AB", t3."R" AS "R", t3."H" AS "H", t3."X2B" AS "X2B", t3."X3B" AS "X3B", t3."HR" AS "HR", t3."RBI" AS "RBI", t3."SB" AS "SB", t3."CS" AS "CS", t3."BB" AS "BB", t3."SO" AS "SO", t3."IBB" AS "IBB", t3."HBP" AS "HBP", t3."SH" AS "SH", t3."SF" AS "SF", t3."GIDP" AS "GIDP", t3."yearID" AS year
    FROM main.batting AS t3) AS t2) AS t0 JOIN main.awards_players AS t1 ON anon_1.year = t1."yearID"
     LIMIT ? OFFSET ?]
    [parameters: (10, 0)]
    (Background on this error at: https://sqlalche.me/e/14/e3q8)

The workaround is to create the column `year` separately and name it outside of `mutate`:

    batting = con.table("batting")
    awards_players = con.table("awards_players")
    year = batting.yearID.name("year")
    left = batting[year, 'RBI']
    right = awards_players
    expr = left.join(right, left.year == right.yearID)
    expr.limit(10).execute()

### PR DESCRIPTION
**Pull request by gforsyth**
Fixes #3529 
Fixes #3515 

I traced the failure mode in #3529 to `ibis.expr.analysis._lift_TableColumn` -- there wasn't a readily apparent way to fix the failure there (or even how to special case it), but removing the entire code path definitely fixes things.  That's not always the best way to fix things, but I think a lot of the expr rewriting stuff is probably causing a lot of issues on the edges.

In any case, I tried a bit to quantify what, if any, time savings are offered by having the `_lift_TableColumn` path available and the results are below.  Usual caveat around how all benchmarks are terrible.

TLDR: For the runs below, removing the expr rewrite has no performance impacts and allows joining on mutated columns, but there may be cases not covered here that would suffer.

**All times are in seconds**

## Running the whole backend test suite 

Use `pytest-repeat` to run each test 3 times to iron out any jitter (maybe)

`pytest -m "<backend>" ibis/backends/tests/ --count=3`

| Backend | with lifting | without lifting | pytest-repeat count |
|---------|--------------|-----------------|---------------------|
| duckdb  | 68.99        | 69.64           | 3                   |
| sqlite  | 83.13        | 84.95           | 3                   |

## Running only tests that hit `_lift_TableColumn`

tests that hit the code in question are:

`test_generic.py::test_select_filter_mutate`
`test_generic.py::test_mutate_rename`
`test_join.py::test_mutating_join[inner | left | right | outer]`
`test_join.py::test_filtering_join[anti | semi ]`

Only join tests (x5)

    pytest ibis/backends/tests/ -m "duckdb" -k "test_filtering_join or test_mutating_join" -v --count=5

| Backend | with lifting | without lifting | pytest-repeat count |
|---------|--------------|-----------------|---------------------|
| duckdb  | 4.15         | 4.15            | 5                   |
| sqlite  | 9.59         | 9.51            | 5                   |
| mysql   | 34.86        | 34.60           | 5                   |

Only generic tests (x5)

``` bash
pytest ibis/backends/tests/test_generic.py -k "test_select_filter_mutate or test_mutate_rename" -v --count=5 -m "duckdb"
```

| Backend | with lifting | without lifting | pytest-repeat count |
|---------|--------------|-----------------|---------------------|
| duckdb  | 3.60         | 3.59            | 15                  |
| sqlite  | 2.90         | 2.88            | 15                  |
| mysql   | 3.62         | 3.58            | 15                  |

## Pathological lifting case?

I tried to come up with a pathological case for where lifting would make a noticeable difference -- if folks have suggestions I'm all ears, but for this case, it makes next to no difference (e.g. the column only gets lifted up a single table-reference, not all the way up to the root table)

``` python
def test_mutate_then_join_no_column_overlap(batting, awards_players):
    left = batting.mutate(year=batting.yearID)
    left = left.mutate(year2=left.year)
    left = left.mutate(year3=left.year2)
    left = left.mutate(year4=left.year3)
    left = left.mutate(year5=left.year4)
    left = left["year5", "RBI"]
    right = awards_players
    expr = left.join(right, left.year5 == right.yearID)
    breakpoint()
    expr.execute()
```

``` diff
diff -u /home/gil/github.com/ibis-project/ibis/ibis/backends/tests/new2 /home/gil/github.com/ibis-project/ibis/ibis/backends/tests/new
--- /home/gil/github.com/ibis-project/ibis/ibis/backends/tests/new2 2022-03-09 14:06:26.810561337 -0500
+++ /home/gil/github.com/ibis-project/ibis/ibis/backends/tests/new  2022-03-09 14:06:27.757568290 -0500
@@ -80,6 +80,6 @@
   predicates:
     boolean = Equals
       left:
-        year5: int64 = r5.year5
+        year5: int64 = r6.year5
       right:
         yearID: int64 = r7.yearID

Diff finished.  Wed Mar  9 14:07:23 2022
```

### CODE REVIEW

**Comment by gforsyth** [comment]
Bonus, this fixed filtering joins on Impala and Clickhouse

**Review by cpcloud — APPROVED** [review]
This is really great. Thank you.
